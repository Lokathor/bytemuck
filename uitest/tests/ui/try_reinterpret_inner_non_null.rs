use bytemuck::TryReinterpretInner;
use core::ptr::NonNull;

fn main() {
  let _: NonNull<[u8; 2]> =
    NonNull::from(&mut 0u8).try_reinterpret_inner().unwrap();
  let _: NonNull<[u16; 2]> =
    NonNull::from(&mut 0u16).try_reinterpret_inner().unwrap();
  let _: NonNull<[u64; 2]> =
    NonNull::from(&mut 0u64).try_reinterpret_inner().unwrap();

  let _: NonNull<[()]> =
    NonNull::from([0u8; 1].as_mut_slice()).try_reinterpret_inner().unwrap();
}
