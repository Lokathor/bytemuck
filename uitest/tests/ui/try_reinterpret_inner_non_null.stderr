error[E0080]: evaluation of `<*const u8 as bytemuck::cast::TryCastRaw<*const [u8; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  |     const ASSERT: () = assert!(
  |  ______________________^
  | |     size_of::<Src>() == size_of::<Dst>(),
  | |     "Attempted conversion between types of different sizes."
  | |   );
  | |___^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:1049:22
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant used
    --> $WORKSPACE/src/cast.rs
     |
     |   const ASSERT: () = <*const Src as TryCastRaw<*const Dst>>::ASSERT;
     |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn <std::ptr::NonNull<u8> as bytemuck::TryReinterpretInner<'_, std::ptr::NonNull<[u8; 2]>>>::try_reinterpret_inner`
 --> tests/ui/try_reinterpret_inner_non_null.rs:6:5
  |
6 |     NonNull::from(&mut 0u8).try_reinterpret_inner().unwrap();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const u16 as bytemuck::cast::TryCastRaw<*const [u16; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  |     const ASSERT: () = assert!(
  |  ______________________^
  | |     size_of::<Src>() == size_of::<Dst>(),
  | |     "Attempted conversion between types of different sizes."
  | |   );
  | |___^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:1049:22
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <std::ptr::NonNull<u16> as bytemuck::TryReinterpretInner<'_, std::ptr::NonNull<[u16; 2]>>>::try_reinterpret_inner`
 --> tests/ui/try_reinterpret_inner_non_null.rs:8:5
  |
8 |     NonNull::from(&mut 0u16).try_reinterpret_inner().unwrap();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<*const u64 as bytemuck::cast::TryCastRaw<*const [u64; 2]>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  |     const ASSERT: () = assert!(
  |  ______________________^
  | |     size_of::<Src>() == size_of::<Dst>(),
  | |     "Attempted conversion between types of different sizes."
  | |   );
  | |___^ the evaluated program panicked at 'Attempted conversion between types of different sizes.', $WORKSPACE/src/cast.rs:1049:22
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn <std::ptr::NonNull<u64> as bytemuck::TryReinterpretInner<'_, std::ptr::NonNull<[u64; 2]>>>::try_reinterpret_inner`
  --> tests/ui/try_reinterpret_inner_non_null.rs:10:5
   |
10 |     NonNull::from(&mut 0u64).try_reinterpret_inner().unwrap();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0080]: evaluation of `<bytemuck::cast::RawSlice<u8> as bytemuck::cast::TryCastRaw<bytemuck::cast::RawSlice<()>>>::ASSERT` failed
 --> $WORKSPACE/src/cast.rs
  |
  |     const ASSERT: () = assert!(
  |  ______________________^
  | |     (size_of::<Dst>() == 0) == (size_of::<Src>() == 0),
  | |     "Attempted conversion between a zero-sized type and a non-zero-sized type"
  | |   );
  | |___^ the evaluated program panicked at 'Attempted conversion between a zero-sized type and a non-zero-sized type', $WORKSPACE/src/cast.rs:1082:22
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` which comes from the expansion of the macro `panic` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant used
    --> $WORKSPACE/src/cast.rs
     |
     |   const ASSERT: () = <RawSlice<Src> as TryCastRaw<RawSlice<Dst>>>::ASSERT;
     |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: the above error was encountered while instantiating `fn <std::ptr::NonNull<[u8]> as bytemuck::TryReinterpretInner<'_, std::ptr::NonNull<[()]>>>::try_reinterpret_inner`
  --> tests/ui/try_reinterpret_inner_non_null.rs:13:5
   |
13 |     NonNull::from([0u8; 1].as_mut_slice()).try_reinterpret_inner().unwrap();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
